common/component/postgresql/awsIAM/aws.go

```diff
@@ -0,0 +1,206 @@
+	// Parse database name from connection string
+	dbName, err := parseDatabaseFromConnectionString(connectionString)
+	if err != nil {
+		return fmt.Errorf("failed to parse database name from connection string to create database %v", err)
+	}
+
+	// check if database exists in connection string
+	var dbExists bool
+	dbExistsCtx, dbExistsCancel := context.WithTimeout(ctx, timeout)
+	err = db.QueryRow(dbExistsCtx, databaseExistsTmpl, dbName).Scan(&dbExists)
+	dbExistsCancel()
+	if err != nil && err != pgx.ErrNoRows {
+		return fmt.Errorf("failed to check if the PostgreSQL database %s exists: %v", dbName, err)
+	}
+
+	// Create database if needed using master password in connection string
+	if !dbExists {
+		createDBCtx, createDBCancel := context.WithTimeout(ctx, timeout)
+		_, err := db.Exec(createDBCtx, fmt.Sprintf(createDatabaseTmpl, dbName))
+		createDBCancel()
+		if err != nil {
+			return fmt.Errorf("failed to create PostgreSQL user: %v", err)
+		}
+	}
+	return nil
+}
+
+func CreateUserAndRoleIfNeeded(ctx context.Context, timeout time.Duration, connectionString string, db pginterfaces.PGXPoolConn) error {
+	// Parse database name from connection string
+	dbName, err := parseDatabaseFromConnectionString(connectionString)
+	if err != nil {
+		return fmt.Errorf("failed to parse database name from connection string to create database %v", err)
+	}
+
+	var userExists bool
+	userExistsCtx, userExistsCancel := context.WithTimeout(ctx, timeout)
+	err = db.QueryRow(userExistsCtx, userExistsTmpl, dbName).Scan(&userExists)
+	userExistsCancel()
+	if err != nil && err != pgx.ErrNoRows {
+		return fmt.Errorf("failed to check if the PostgreSQL user %s exists: %v", dbName, err)
+	}
+
+	// Create the user if it doesn't exist
+	if !userExists {
+		_, err = db.Exec(ctx, fmt.Sprintf(createUserTmpl, dbName))
+		if err != nil {
+			return fmt.Errorf("failed to create PostgreSQL user: %v", err)
+		}
+	}
+
+	// Check if the role is already granted
+	var roleGranted bool
+	roleGrantedCtx, roleGrantedCancel := context.WithTimeout(ctx, timeout)
+	err = db.QueryRow(roleGrantedCtx, checkRoleTmpl, dbName).Scan(&roleGranted)
+	roleGrantedCancel()
+	if err != nil && err != pgx.ErrNoRows {
+		return fmt.Errorf("failed to check if the role %v is already granted to the PostgreSQL user %s: %v", awsRole, dbName, err)
+	}
+
+	// Grant the role if it's not already granted
+	if !roleGranted {
+		grantRoleCtx, grantRoleCancel := context.WithTimeout(ctx, timeout)
+		_, err := db.Exec(grantRoleCtx, fmt.Sprintf(grantRoleTmpl, awsRole, dbName))
+		grantRoleCancel()
+		if err != nil {
+			return fmt.Errorf("failed to grant PostgreSQL user role: %v", err)
+		}
+	}
+	return nil
+}
+
+func InitAWSDatabase(ctx context.Context, config *pgxpool.Config, db pginterfaces.PGXPoolConn, timeout time.Duration, connString string, awsAccessKey, awsSecretKey string) error {
+	err := CreateDatabaseIfNeeded(ctx, timeout, connString, db)
```

Comment:
can this create steps be opt in or opt out? so the component can be configured with a connectionString without password and then the logic here would just automatically provide the aws token for the user in the connection string

Reply:
ohh, yes I can account for that. thank you for the feedback!

Reply:
I've tried to get this to work without using the master username/password. However, I haven't been able to get it to work this way. I started by creating my IAM user, creating the same user name in my database and granting the rds_iam permissions within the specific database from the connection string. However, I keep seeing the `FATAL: PAM authentication failed for user <username>"` even after verifying that the user has the rds_iam role granted to it within the proper database. Not sure what to try next here. I even tried following this tutorial too to see if I missed anything and I don't think that I did: https://www.linkedin.com/pulse/authenticating-postgres-rds-logins-using-aws-iam-cross-devadoss/